phases/16-multi-agent-and-swarms/19-swarm-optimization-pso-aco/code/main.py

```
@@ -0,0 +1,188 @@
+    print(f"  final g_best={history[-1]:.4f} (optimum = 1.0000 at (0.72, 0.40))")
+
+    print("\n" + "=" * 72)
+    print("AMRO-S — 200 tasks routed across 3 agents × 4 task-types")
+    print("=" * 72)
+    rand_quality, aco_quality = run_amro_s()
+    print(f"  random routing avg quality: {rand_quality:.3f}")
+    print(f"  ACO routing    avg quality: {aco_quality:.3f}")
+    print(f"  improvement: {(aco_quality - rand_quality) / rand_quality * 100:+.1f}%")
+
+    # Show final pheromone table for inspection
+    rng = random.Random(0)
```

Ruff (0.15.11)

[error] Standard pseudo-random generators are not suitable for cryptographic purposes

(S311)